fix: changing SceneEventData allocator from TmpJob to Persistent

NoelStephensUnity opened the pull request:
[MTT-1320](https://jira.unity3d.com/browse/MTT-1320)
This fixes the error where you can potentially see:
"JobTempAlloc has allocations that are more than 4 frames old - this is not allowed and likely a leak"
Due to the fact that SceneEventData was using the temporary job memory allocator as opposed to persistent ( scene loading will normally take more than 4 frames).

The included test validates no errors occur when the FastBufferReader is held for longer than 4 frames.

NoelStephensUnity committed:
fix

this removes the memory leak warning where the InternalBuffer reader was being created with Allocator.TmpJob (if around for more than 4 frames it generates a warning).  Switch this to be Allocator.Persistent since loading and scene synchronization will most likely happen over more than a 4 frame period of time.

NoelStephensUnity committed:
test

first half.

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
test

Backing out scenes loaded for MultiSceneSynchronizationTest

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -475,7 +475,8 @@ internal void Deserialize(FastBufferReader reader)
                             m_HasInternalBuffer = true;
-                            InternalBuffer = new FastBufferReader(reader.GetUnsafePtrAtCurrentPosition(), Allocator.TempJob, reader.Length - reader.Position);
+                            // We use Allocator.Persistent since scene loading could take longer than 4 frames
+                            InternalBuffer = new FastBufferReader(reader.GetUnsafePtrAtCurrentPosition(), Allocator.Persistent, reader.Length - reader.Position);
```
And I'm assuming we want and need it to live that extended period?

NoelStephensUnity replied:
Yeah.   Allocator.TempJob will start warning if you hold onto that for longer than 4 frames.  That buffer is to be used for scene loading which could take dozens to hundreds of frames depending upon the size of the scene being loaded.

LukeStampfli reviewed (approved):
Yes this change makes a lot of sense. Ship it! 🚀

NoelStephensUnity committed:
fix and test

While writing the test for this PR I found an issue where it could try to dispose SceneEventData instance more than once during a MultiInstance unit test.  Removing the EndSceneEvent(sceneEventId); from line 998 fixed this issue.

The rest of this commit finalizes the test for this PR.

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
style

replaced UNITY_EDITOR || DEVELOPMENT_BUILD with UNITY_INCLUDE_TESTS

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
refactor

Created a better test to verify that changing the FastBufferReader's memory allocation type from Allocator.TmpJob to Allocator.Persistent fixes the issue where a scene event could take longer than 4 frames and cause an error and log that a potential memory leak could be happening.

Removed previous test and associated changes to NetworkSceneManager.

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent

NoelStephensUnity committed:
refactor

missed one #if UNITY_INCLUDE_TESTS

NoelStephensUnity committed:
Merge branch 'develop' into fix/SceneEventData-Allocator-TmpJob-to-Persistent
